src/GafferScene/Instancer.cpp

```diff
@@ -766,11 +1255,49 @@ void Instancer::hash( const Gaffer::ValuePlug *output, const Gaffer::Context *co
 		scalePlug()->hash( h );
 		attributesPlug()->hash( h );
 		attributePrefixPlug()->hash( h );
+
+		seedEnablePlug()->hash( h );
+		seedVariablePlug()->hash( h );
+		seedPossibilitiesPlug()->hash( h );
+		seedPermutationPlug()->hash( h );
+		rawSeedPlug()->hash( h );
+
+		for( ContextVariablePlug::Iterator it( contextVariablesPlug() ); !it.done(); ++it )
+		{
+			const ContextVariablePlug *plug = it->get();
+			if( plug->enabledPlug()->getValue() )
+			{
+				plug->namePlug()->hash( h );
+				plug->quantizePlug()->hash( h );
+			}
+		}
+
+		if( timeOffsetPlug()->enabledPlug()->getValue() )
+		{
+			timeOffsetPlug()->namePlug()->hash( h );
+			timeOffsetPlug()->quantizePlug()->hash( h );
+		}
 	}
 	else if( output == prototypeChildNamesPlug() )
 	{
 		enginePlug()->hash( h );
 	}
+	else if( output == variationsPlug() )
+	{
+		// The sum of the variations across different engines depends on all the engines, but
+		// not their order.  We can create a cheap order-independent hash by summing the hashes
+		// all of the engines
+		std::atomic<uint64_t> h1Accum( 0 ), h2Accum( 0 );
+		std::function<bool(const GafferScene::ScenePlug *, const GafferScene::ScenePlug::ScenePath &)> functor =[this, &h1Accum, &h2Accum]( const GafferScene::ScenePlug *scene, const GafferScene::ScenePlug::ScenePath &path )
+		{
+			IECore::MurmurHash h = enginePlug()->hash();
+			h1Accum.fetch_add( h.h1(), std::memory_order_relaxed );
```

Comment:
Memory ordering makes my head explode! Two questions :

- Is `memory_order_relaxed` on its own enough to guarantee that the read at line 1299 sees the final value?
- Is this operation significant enough that we should worry about relaxed order vs a good old `+=`?

Reply:
Oh, I wasn't trying to get fancy, I just missed that std::atomic actually had a += that was guaranteed atomic, so I just glanced at the docs for fetch_add (https://en.cppreference.com/w/cpp/atomic/atomic/fetch_add), and saw that the example on that page was this exact circumstance ( multiple threads try to increment at once, then read in the main thread once separate threads are finished ), and the docs example used memory_order_relaxed, so I assumed it was adequate in this case.  It's curious that the `+=` is hardcoded to use memory_order_seq_cst.

In this case, having thought about it a bit, memory_order_relaxed is probably correct ( the order of atomics written by worker threads doesn't matter at all, and the docs seem to say pretty clearly that even with relaxed, the correct value will be seen after worker threads finish, even though I don't really understand this stuff well enough to argue myself why that would be guaranteed ).  But this is completely unimportant from a performance point of view in this case ( the only way it could possibly come up would be if you instanced onto a million separate pointclouds, each with 1 point, and that's not reasnoable ), so I've just switched to += to look simpler.